lib/SILOptimizer/Mandatory/ClosureLifetimeFixup.cpp

```diff
@@ -363,6 +363,61 @@ static SILValue insertMarkDependenceForCapturedArguments(PartialApplyInst *pai,
   return curr;
 }
 
+/// Returns the (single) "endAsyncLet" builtin if \p startAsyncLet is a
+/// "startAsyncLet" builtin.
+static BuiltinInst *getEndAsyncLet(BuiltinInst *startAsyncLet) {
+  if (startAsyncLet->getBuiltinKind() != BuiltinValueKind::StartAsyncLet)
+    return nullptr;
+
+  BuiltinInst *endAsyncLet = nullptr;
+  for (Operand *op : startAsyncLet->getUses()) {
+    auto *endBI = dyn_cast<BuiltinInst>(op->getUser());
+    if (endBI && endBI->getBuiltinKind() == BuiltinValueKind::EndAsyncLet) {
+      // At this stage of the pipeline, it's always the case that a
+      // startAsyncLet has an endAsyncLet: that's how SILGen generates it.
+      // Just to be on the safe side, do this check.
+      if (endAsyncLet)
+        return nullptr;
+      endAsyncLet = endBI;
```

Comment:
Somewhat tangential, this made me wonder whether we mark this Builtin as no duplicatable via the usual means? We should if we don't.

Reply:
Hm, why would cloning this builtin not work?

Reply:
Do we run `ClosureLifetimeFixup` before any optimization that would clone the builtin that could prevent the optimized fixup?

Reply:
Either way should be no problem. If it's cloned before (which I'm pretty sure it's not the case), then ClosureLifetimeFixup will extend the life ranges to the end of the function: a conservative default.

If it's cloned afterwards, then I also see no problem.

Reply:
I think it is preferable to not extend the lifetime over not been able to duplicate (especially since we can limit that to the pre closure fixup form). But if we don't run any cfg optimizing changes in mandatory opts before closure lifetime mixup then it is not an issue.